lib/getWorkflow.js

```diff
@@ -158,23 +250,51 @@ const calculateEdges = async (jobs, triggerFactory, externalDownstreamOrs, exter
     await Promise.all(
         Object.keys(jobs).map(async jobName => {
             const job = jobs[jobName];
+            const stageObj = job.stage;
 
             if (Array.isArray(job.requires)) {
                 // Calculate which upstream jobs trigger the current job
-                const upstreamOr = new Set(job.requires.filter(name => name.charAt(0) === '~'));
-                const upstreamAnd = new Set(job.requires.filter(name => name.charAt(0) !== '~'));
+                const upstreamOr = new Set(
+                    job.requires.filter(name => name.charAt(0) === '~' || name.startsWith(STAGE_PREFIX))
+                );
+                const upstreamAnd = new Set(
+                    job.requires.filter(name => name.charAt(0) !== '~' && !name.startsWith(STAGE_PREFIX))
+                );
                 const isJoin = upstreamAnd.size > 1;
 
                 upstreamOr.forEach(src => {
-                    edges.push({ src: filterNodeName(src), dest: jobName });
+                    const [source] = src.split(':');
+
+                    if (stageObj) {
+                        let stageEdge = { src: filterNodeName(src), dest: jobName };
+
+                        // Remove stage prefix when src is not stage setup
+                        if (!stageObj.startFrom) {
+                            stageEdge = { src: filterNodeName(source), dest: jobName };
+                        }
```

Comment:
This block may not be needed. What is the use case which needs this?

I believe when the condition satisfies, `src` and `source` would be identical.
```suggestion
                        // Remove stage prefix when src is not stage setup
                        if (!stageObj.startFrom) {
                            stageEdge = { src: filterNodeName(source), dest: jobName };
                        }
```

Reply:
I think there's redundant edges if we don't have this

Reply:
removed